### PR TITLE
[Sprint 4] Feature add book logic 3

### PR DESCRIPTION
**Pull request by nguyenminhduong**
Redmine https://dev.framgia.com/redmine/issues/136395
Feature add book logic

### CODE REVIEW

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -63,4 +63,8 @@
     <string name="select_image_from_gallery_or_camera">Select image from gallery or camera</string>
     <string name="select_office">Select office</string>
     <string name="are_u_want_to_sign_out">Are you sure want to sign out?</string>
+    <string name="select_category">Select category</string>
```
Comment (trainer-android-fr) [inline app/src/main/res/values/strings.xml]: FATAL: [MissingTranslation] "`select_category`" is not translated in "vi" (Vietnamese)
 If an application has more than one locale, then all the strings declared in one language should also be translated in all other languages.

If the string should *not* be translated, you can add the attribute `translatable="false"` on the `<string>` element, or you can define all your non-translatable strings in a resource file called `donottranslate.xml`. Or, you can ignore the issue with a `tools:ignore="MissingTranslation"` attribute.

By default this detector allows regions of a language to just provide a subset of the strings and fall back to the standard language strings. You can require all regions to provide a full translation by setting the environment variable `ANDROID_LINT_COMPLETE_REGIONS`.

You can tell lint (and other tools) which language is the default language in your `res/values/` folder by specifying `tools:locale="languageCode"` for the root `<resources>` element in your resource file. (The `tools` prefix refers to the namespace declaration `http://schemas.android.com/tools`.)

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -63,4 +63,8 @@
...
+
+    <!-- date format -->
+    <string name="date_format_yyyy_MM_dd">yyyy-MM-dd</string>
```
Comment (trainer-android-fr) [inline app/src/main/res/values/strings.xml]: FATAL: [MissingTranslation] "`date_format_yyyy_MM_dd`" is not translated in "vi" (Vietnamese)
 If an application has more than one locale, then all the strings declared in one language should also be translated in all other languages.

If the string should *not* be translated, you can add the attribute `translatable="false"` on the `<string>` element, or you can define all your non-translatable strings in a resource file called `donottranslate.xml`. Or, you can ignore the issue with a `tools:ignore="MissingTranslation"` attribute.

By default this detector allows regions of a language to just provide a subset of the strings and fall back to the standard language strings. You can require all regions to provide a full translation by setting the environment variable `ANDROID_LINT_COMPLETE_REGIONS`.

You can tell lint (and other tools) which language is the default language in your `res/values/` folder by specifying `tools:locale="languageCode"` for the root `<resources>` element in your resource file. (The `tools` prefix refers to the namespace declaration `http://schemas.android.com/tools`.)

File: app/src/main/res/values/strings.xml (hunk 1 above)
Comment (trainer-android-fr) [inline app/src/main/res/values/strings.xml]: WARNING: [UnusedResources] The resource `R.string.select_category` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/res/values/strings.xml (hunk 2 above)
Comment (trainer-android-fr) [inline app/src/main/res/values/strings.xml]: WARNING: [UnusedResources] The resource `R.string.date_format_yyyy_MM_dd` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/java/com/framgia/fbook/widget/dialog/DialogManagerImpl.kt [hunk 3]
```diff
@@ -133,4 +141,22 @@ open class DialogManagerImpl(private val mContext: Context) : DialogManager {
         .onPositive { materialDialog, _ -> materialDialog.dismiss() }
         .show()
   }
+
+  override fun dialogDatePicker(
+      onDateSetListener: DatePickerDialog.OnDateSetListener, calendar: Calendar): DialogManager {
+    mDatePickerDialog = DatePickerDialog(mContext, onDateSetListener, calendar.get(Calendar.YEAR),
+        calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH))
+    mDatePickerDialog!!.setButton(DialogInterface.BUTTON_NEUTRAL,
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/widget/dialog/DialogManagerImpl.kt]: avoid !! in kotlin it is common cause raise Crash, If really necessary We should check !=null before call

File: app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt [hunk 4]
```diff
@@ -1,29 +1,34 @@
...
 import com.framgia.fbook.databinding.ActivitySharebookBinding
 import com.framgia.fbook.screen.BaseActivity
+import com.framgia.fbook.utils.common.DateTimeUtils
 import com.framgia.fbook.utils.navigator.Navigator
 import com.fstyle.library.MaterialDialog
 import com.fstyle.structure_android.widget.dialog.DialogManager
 import pl.aprilapps.easyphotopicker.DefaultCallback
 import pl.aprilapps.easyphotopicker.EasyImage
 import java.io.File
+import java.util.*
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt]: avoid import like that.
Pls check coding convention of Framgia [link](https://github.com/framgia/coding-standards/blob/master/vn/android/codingstyleguide.md#14-fully-qualify-imports)

File: app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt [hunk 5]
```diff
@@ -34,12 +39,16 @@ class ShareBookActivity : BaseActivity(), ShareBookContract.ViewModel, ItemImage
...
+  private val mListCategory = mutableListOf<Category>()
 
   val titleError: ObservableField<String> = ObservableField()
   val authorError: ObservableField<String> = ObservableField()
   val descriptionError: ObservableField<String> = ObservableField()
-  var currentOffice: ObservableField<String> = ObservableField()
+  val publishDate: ObservableField<String> = ObservableField()
+  val currentOffice: ObservableField<String> = ObservableField()
   private var currentOfficePosition: Int = 0
+  val currentCategory: ObservableField<String> = ObservableField()
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt]: pls move this line above line [#49](https://github.com/framgia/fbook_android/pull/48/files#diff-9a72c6cadf753e6fb58db23f5481999aR49)

File: app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt [hunk 6]
```diff
@@ -111,14 +121,45 @@ class ShareBookActivity : BaseActivity(), ShareBookContract.ViewModel, ItemImage
     updateCurrentOffice(currentOfficePosition)
   }
 
+  override fun onGetCategorySuccess(listCategory: List<Category>?) {
+    mListCategory.addAll(listCategory!!)
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt]: avoid !!

File: app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt [hunk 7]
```diff
@@ -111,14 +121,45 @@ class ShareBookActivity : BaseActivity(), ShareBookContract.ViewModel, ItemImage
...
+    updateCurrentCategory(currentCategoryPosition)
+  }
+
+  override fun onDateSet(datePicker: DatePicker?, year: Int, month: Int, day: Int) {
+    if (year == 0) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt]: pls add comment declare when the year is zero above this line

File: app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt [hunk 8]
```diff
@@ -111,14 +121,45 @@ class ShareBookActivity : BaseActivity(), ShareBookContract.ViewModel, ItemImage
...
+      updatePublishDate("")
+      return
+    }
+    val date = DateTimeUtils.getDateStringFromDayMonthYear(year, month, day,
+        getString(R.string.date_format_yyyy_MM_dd))
+    updatePublishDate(date!!)
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt]: avoid !!

File: app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt [hunk 9]
```diff
@@ -111,14 +121,45 @@ class ShareBookActivity : BaseActivity(), ShareBookContract.ViewModel, ItemImage
...
+  }
+
   override fun onError(baseException: BaseException) {
   }
 
   private fun updateCurrentOffice(position: Int) {
+    if (mListOffice.size <= 0) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt]: change to ``mListOffice.isEmpty``

File: app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt [hunk 10]
```diff
@@ -111,14 +121,45 @@ class ShareBookActivity : BaseActivity(), ShareBookContract.ViewModel, ItemImage
...
+      return
+    }
     currentOffice.set(mListOffice[position].name)
     mBookRequest.officeId = mListOffice[position].id
   }
 
+  private fun updateCurrentCategory(position: Int) {
+    if (mListCategory.size <= 0) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt]: like above

File: app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt [hunk 11]
```diff
@@ -146,17 +187,30 @@ class ShareBookActivity : BaseActivity(), ShareBookContract.ViewModel, ItemImage
...
+    mDialogManager.dialogListSingleChoice(getString(R.string.select_office), workSpace,
+        currentOfficePosition,
+        MaterialDialog.ListCallbackSingleChoice { _, _, position, _ ->
+          currentOfficePosition = position
+          updateCurrentOffice(currentOfficePosition)
+          true
+        })
+  }
+
+  fun onPickCategory() {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/screen/sharebook/ShareBookActivity.kt]: pls rearrange methods in this file follow framgia coding convention
method with modify ``public`` at top of file, next default, end of file is location for method with modify ``private``

File: app/src/main/java/com/framgia/fbook/widget/dialog/DialogManagerImpl.kt [hunk 12]
```diff
@@ -1,17 +1,24 @@
...
+import java.util.*
 
 /**
  * Created by le.quang.dao on 14/03/2017.
  */
 
 open class DialogManagerImpl(private val mContext: Context) : DialogManager {
   private var mProgressDialog: MaterialDialog? = null
+  private var mDatePickerDialog: DatePickerDialog? = null
+  private var mDatePicker: DatePicker? = null
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/fbook/widget/dialog/DialogManagerImpl.kt]: pls move this variable inside method use it

**Pull request merged**